File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
const letterCombinations = (digits) => {
  const res = [];
  const dict = [,, 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz'];
  const build = (prefix, digits) => {
    const str = dict[digits % 10];
    digits = Math.floor(digits / 10);
    for (const char of str) {
      digits ? build(char + prefix, digits) : res.push(char + prefix);
    }
  }
  digits && build('', digits);
  return res;
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.go
var dict = []string{"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"}

func build(ans *[]string, rns *[]rune, prefix string, i int) {
  for _, rn := range dict[(*rns)[i] - '0'] {
    if i - 1 >= 0 {
      build(ans, rns, string(rn) + prefix, i - 1)
    } else {
      *ans = append(*ans, string(rn) + prefix)
    }
  }
}

func letterCombinations(digits string) []string {
  ans := []string{}
  if len(digits) > 0 {
    rns := []rune(digits)
    build(&ans, &rns, "", len(digits) - 1)
  }
  return ans
}